Restore previous test run state

laszlourszuly opened the pull request:
Fixes #13

Commit:
Allow git to handle default config file

Commit:
Persist and recover current state

Commit:
Restore last dumped metrics

Commit:
Fix test state publishing

Commit:
Disable broken tests

These will be fixed in separate PR. Disabling for now so that the
test project can be built properly at least.

blenessy commented on explorer/service/src/main/java/org/radixdlt/explorer/metrics/MetricsProvider.java:
```diff
@@ -178,25 +170,62 @@ private void calculateMetrics(Map<String, SystemInfo> info) {
             averageTps = Math.round(progress / seconds);
             peakTps = Math.max(peakTps, tps);
             calculatedMetrics = new Metrics(tps, progress, averageTps, peakTps);
+            dumpCurrentMetrics();
```
Not good idea to do I/O while holding a lock (hint: I/O operations take milliseconds to complete)

I think an asynchronous (periodically scheduled) task is a better design for dumpingCurrentMetrics.
A caveat is to only execute the Task if the test is in RUNNING state.

laszlourszuly replied:
I fear moving the "dump-to-file" operation outside the synchronized context may force us into magical race conditions where we're not writing what we just calculated (this method may be called at any time, from any thread).

Any further input on how to prevent these race conditions from a separate worker thread would be greatly appreciated.

blenessy replied:
If you do all the writing from a single background thread then there is no need to synchronize dumpingCurrentMetrics() ?

laszlourszuly replied:
But I can't guarantee that it's one single background thread. It depends on the OkHttp client's thread pool (>1 background threads) and the load it currently experiences.

laszlourszuly replied:
Ah! But wait! You, of course, mean that I should "enqueue" a new "dump task" and let that queue be consumed from a single background thread.

Now it makes sense! Give me a coupe of minutes to fix this! :)

laszlourszuly replied:
Now fixed!

blenessy replied:
Sorry for being unclear. Yes, you got my point correctly you serialised writes to the file and thus prevented potential race conditions.

blenessy commented on explorer/service/src/main/java/org/radixdlt/explorer/metrics/MetricsProvider.java:
```diff
@@ -178,25 +170,62 @@ private void calculateMetrics(Map<String, SystemInfo> info) {
+            String line = calculatedMetrics.toString();
+            try {
                 byte[] data = line.getBytes(UTF_8);
                 Files.write(metricsDumpPath, data, CREATE, WRITE, APPEND);
```
Only a good idea if you know the maximum size that the file can grow to.

laszlourszuly replied:
Hm... I'm not sure I follow. The file is cleared every time the test starts (enters the `STARTED` state). From there the file size grows linearly during the execution time of the test. Each line has fixed size so I can make a coarse grained estimate of the terminal dump file size. But I'm not sure I see what to do with this information.

blenessy replied:
Yup 3d2adaa is the logic I was looking for. Nice O(1) algo.

blenessy commented on explorer/service/src/main/java/org/radixdlt/explorer/metrics/MetricsProvider.java:
```diff
@@ -178,25 +170,62 @@ private void calculateMetrics(Map<String, SystemInfo> info) {
+    private void restoreMetrics() {
+        if (metricsDumpPath != null) {
+            try {
+                List<String> lines = Files.readAllLines(metricsDumpPath, UTF_8);
```
Only a good idea if you know for sure that the max file size **always** fits nicely in memory.

laszlourszuly replied:
Good point. I don't expect the file to grow "in absurdum" though as it's reset every time a test run starts.

laszlourszuly replied:
I have fixed this now (using a `RandomAccessFile` and reading from the end instead).

blenessy reviewed (changes requested):
The handling of **dumpCurrentMetrics** seems dangerous .

laszlourszuly commented:
Thanks for feedback @blenessy! Let's have a "offline" chat regarding the "dump to file" alternatives.

Commit:
Only read last line from metrics dump file

The previous implementation read the entire metrics dump file into
memory and picked the last line from there. This change will start
reading from the end of the file and only read the last line which
should utilise resources more efficiently.

Commit:
Enqueue dump tasks rather than blocking execution
